decomposition_tensor.py

```diff
@@ -0,0 +1,162 @@
+
+from functools import wraps
+import torch
+from torch._decomp import decomposition_table
+import torch.nn.functional as F
+from torch.utils._pytree import tree_map
+from torch.testing._internal.logging_tensor import LoggingTensor, capture_logs
+from torch.testing._internal.common_utils import run_tests, TestCase
+
+# Goals:
+# - we want something reusable that can compose with any subclass
+# Non-goal:
+# - should work with both __torch_dispatch__ and __torch_function__
+#   - decomposition table is aten to aten (alternatively, parametrize on decomposition table?)
+#
+# Should this be a wrapper subclass ("has-a") or just use inheritance? Both are implemented below.
+# - Wrapper subclass API `DecompositionTensor(LoggingTensor(t))`
+#   - will need to redispatch
+# - OR provide helper dynamically inherit from the provided subclass and just override its __torch_dispatch__
+#   function to the decorated version
+#   + won't need to redispatch
+# - OR User just writes inline if necessary
+#   + more customizability
+#
+# How does this work with torch function?
+# - It doesn't. torch function will try to rewrap the output again and error if we don't disable it
+#
+# What is progressive lowering tensor and how does this compare?
+# - TODO
+
+aten = torch.ops.aten
+
+skip_list = [aten.add.Tensor, aten.to.dtype, aten.div.Tensor, aten.clamp.default]
+
+# 1) Wrapper subclass inline version
+class DecompositionTensor(torch.Tensor):
+    @staticmethod
+    def __new__(cls, e):
+        r = torch.Tensor._make_wrapper_subclass(cls, e.shape, dtype=e.dtype, requires_grad=False)
+        r.elem = e
+        return r
+
+    # We may be able to remove this line in the future when Ed's PR lands
+    __torch_function__ = torch._C._disabled_torch_function_impl
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):  # type: ignore
+        def unwrap(e):
+            return e.elem if isinstance(e, DecompositionTensor) else e
+
+        def wrap(e):
+            return DecompositionTensor(e) if isinstance(e, torch.Tensor) else e
+
+        if func in skip_list:
+            # Check skip-list first, so "backend" has a chance to see non-decomposed ops
+            return tree_map(wrap, func(*tree_map(unwrap, args), **tree_map(unwrap, kwargs)))
+        elif func in decomposition_table:
+            return decomposition_table[func](*args, **kwargs)
+        else:
+            raise NotImplementedError(f"{func.__name__} does not have a decomposition and is not in skip_list")
+
+# 2) Decorator (Alban's suggestion)
+def decompose(skip_list, missing_ops=None):
+    def _decompose(f):
+        @wraps(f)
+        def wrapper(cls, func, types, args=(), kwargs=None):
+            if func in skip_list:
+                # Functions that the layers below are able to handle
+                return f(cls, func, types, args, kwargs)
```

Comment:
how come unwrapping isn't needed in this version?

Reply:
Ahh `f` is the `__torch_dispatch__` function not the aten op, so the unwrapping will still happen there. Maybe I should rename it to something better so that is clearer...